components/util/model.jsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import { immutableMerge } from './index.js';
+
+export function Model({ initialState, methods }) {
+  const listeners = [];
+  const notifyListeners = nextState => listeners.forEach(fn => fn(nextState));
+
+  const model = {
+    state: Object.freeze(initialState),
+    subscribe: fn => listeners.push(fn),
+    unsubscribe: fn => listeners.splice(listeners.indexOf(fn), 1),
+  };
+
+  Object.keys(methods).forEach((method) => {
+    model[method] = (...args) => {
+      const updates = methods[method].apply(model, args);
+      if (updates && typeof updates === 'object') {
+        model.state = immutableMerge(model.state, updates);
+        notifyListeners(model.state);
+      }
+      // could also check here to see if fn returns a promise and then wait
+      // until it resolves to merge and notifyListeners. that would make
+      // async methods really nice to work with.
+      // ...should this return  something? maybe return `updates`?
```

Comment:
No edge cases come to mind. I just didn't implement it since I don't have a use for it.

The async stuff would be useful whenever you want to update the model based on, say, a `fetch`. Following the example jumpstate used:

```js
const myModel = Model({
  initialState: {
    count: 0
  },
  methods: {
    inc() {
      // whatever is returned in a method gets merged into the state object
      return { count: this.state.count + 1 };
    },
    dec() {
      return { count: this.state.count - 1 };
    },
    asyncInc() {
      return new Promise((resolve) => {
        setTimeout(() => resolve({ count: this.state.count + 1 }) , 2000);
      });
    }
  }
});
```

A more practical example would be:

```js
const myModel = Model({
  initialState: {
    products: [],
    isLoaded: false,
    searchQuery: '',
    errorMessage: '',
  },
  methods: {
    search(searchQuery) {
      return { searchQuery };
    },
    loadProducts() {
      return fetch('/products.json')
        .then((products) => ({ isLoaded: true, products }))
        .catch(() => ({ isLoaded: true, errorMessage: 'Failed to fetch products' }))
    }
  }
});
```